### PR TITLE
TSK-1170 Ngxs store

### PR DESCRIPTION
**Pull request by sofie29**

### CODE REVIEW

File: web/src/app/administration/workbasket/details/information/workbasket-information.component.spec.ts [hunk 1]
```diff
@@ -113,7 +123,7 @@ describe('WorkbasketInformationComponent', () => {
     expect(component.workbasket.workbasketId).toEqual(component.workbasketClone.workbasketId);
   });
 
-  it('should reset requestInProgress after saving request is done', () => {
+  it('should reset requestInProgress after saving request is done', async(() => {
```
Comment (mustaphazorgati) [inline web/src/app/administration/workbasket/details/information/workbasket-information.component.spec.ts:126]: why did you change this line?

File: web/src/app/shared/classification-types-selector/classification-types-selector.component.spec.ts [hunk 2]
```diff
@@ -8,9 +9,10 @@ describe('ClassificationTypesSelectorComponent', () => {
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
-      declarations: [ClassificationTypesSelectorComponent]
-    })
-      .compileComponents();
+      imports: [NgxsModule.forRoot()],
```
Comment (mustaphazorgati) [inline web/src/app/shared/classification-types-selector/classification-types-selector.component.spec.ts:12]: no store mock?

File: web/src/app/shared/classification-types-selector/classification-types-selector.component.ts [hunk 3]
```diff
@@ -1,28 +1,30 @@
...
+  constructor(private store: Store) {}
+
+  ngOnInit(): void {
   }
 
-  select(value: string) {
-    this.classificationTypeSelected = value;
+  select(value: string): void {
+    this.store.dispatch(new SetSelectedClassificationType(value));
     this.classificationTypeChanged.emit(value);
```
Comment (mustaphazorgati) [inline web/src/app/shared/classification-types-selector/classification-types-selector.component.ts:26]: do we still need output this?

File: web/src/app/shared/services/classifications/classification-categories.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  it('should return customisation with inserted missing icon', async(() => {
+    const expectedCustomisation: Customisation = { EN:
+      { classifications:
+        { categories:
+          { missing: 'assets/icons/categories/missing-icon.svg' } } } };
+
+    const initialCustomisations: Customisation[] = [
+      { EN: { classifications: { categories: {} } } },
+      { EN: { classifications: {} } },
+      { EN: {} }
```
Comment (mustaphazorgati) [inline web/src/app/shared/services/classifications/classification-categories.service.spec.ts:29]: I am missing a test case with 2 languages.

File: web/src/app/shared/services/classifications/classification-categories.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    categoryService = TestBed.get(ClassificationCategoriesService);
+    httpMock = TestBed.get(HttpTestingController);
+  });
+
+  it('should return customisation with inserted missing icon', async(() => {
+    const expectedCustomisation: Customisation = { EN:
+      { classifications:
+        { categories:
+          { missing: 'assets/icons/categories/missing-icon.svg' } } } };
```
Comment (mustaphazorgati) [inline web/src/app/shared/services/classifications/classification-categories.service.spec.ts:24]: can we make this a constant in the service and just import that constant here?

File: web/src/app/shared/services/classifications/classification-categories.service.ts [hunk 6]
```diff
@@ -2,87 +2,36 @@ import { HttpClient } from '@angular/common/http';
...
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { Customisation } from '../../../models/customisation';
+
+const customisationUrl = 'environments/data-sources/taskana-customization.json';
+
+export interface CategoriesResponse { [key: string]: string[] }
 
 @Injectable()
 export class ClassificationCategoriesService {
```
Comment (mustaphazorgati) [inline web/src/app/shared/services/classifications/classification-categories.service.ts:14]: This name is not correct anymore. We should rename it and think about a way to restructure our services

File: web/src/app/shared/services/classifications/classification-categories.service.ts [hunk 7]
```diff
@@ -2,87 +2,36 @@ import { HttpClient } from '@angular/common/http';
...
-  }
-
-  private getCustomCategoriesObject(categories: Array<string>): Object {
-    return this.customFieldsService.getCustomObject(
-      this.getDefaultCategoryMap(categories), 'classifications.categories'
+  getCustomisation(): Observable<Customisation> {
+    return this.httpClient.get<Customisation>(customisationUrl).pipe(
+      map(customisation => {
+        Object.keys(customisation).forEach(lang => {
+          if (!customisation[lang].classifications) customisation[lang].classifications = {};
```
Comment (mustaphazorgati) [inline web/src/app/shared/services/classifications/classification-categories.service.ts:29]: Please use https://github.com/jonschlinkert/set-value

File: web/src/app/shared/services/classifications/classifications.service.ts [hunk 8]
```diff
@@ -21,13 +22,13 @@ export class ClassificationsService {
   private classificationSaved = new Subject<number>();
   private classificationResourcePromise: Promise<ClassificationResource>;
   private lastDomain: string;
+  // TODO: this should not be here in the service
```
Comment (mustaphazorgati) [inline web/src/app/shared/services/classifications/classifications.service.ts:25]: We should do some research and think about this in general.
Do we want select statements within the service? If not, what are the alternatives.

File: web/src/app/shared/services/classifications/classifications.service.ts [hunk 9]
```diff
@@ -67,9 +68,12 @@ export class ClassificationsService {
   getClassification(id: string): Promise<ClassificationDefinition> {
     return this.httpClient.get<ClassificationDefinition>(`${this.url}${id}`)
       .pipe(tap((classification: ClassificationDefinition) => {
-        if (classification) {
-          this.classificationCategoriesService.selectClassificationType(classification.type);
-        }
+        // why do we need that?
```
Comment (mustaphazorgati) [inline web/src/app/shared/services/classifications/classifications.service.ts:71]: I don't know. Lets find this out together.

File: web/src/app/shared/services/classifications/classifications.service.ts [hunk 10]
```diff
@@ -108,10 +112,9 @@ export class ClassificationsService {
   // #endregion
 
   private getClassificationObservable(classificationRef: Observable<any>): Observable<Array<Classification>> {
-    const classificationTypes: Observable<string> = this.classificationCategoriesService.getSelectedClassificationType();
     return combineLatest(
       [classificationRef,
-        classificationTypes]
+        this.classificationTypeSelected$]
```
Comment (mustaphazorgati) [inline web/src/app/shared/services/classifications/classifications.service.ts:117]: We can simplify this.

File: web/src/app/store/Classification-state.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
+import { Action, Selector, State, StateContext } from '@ngxs/store';
```
Comment (mustaphazorgati) [inline web/src/app/store/Classification-state.ts:1]: This state is still one huge file.
Lets talk about how we can handle states which will get bigger.

File: web/src/app/store/Classification-state.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+import { Observable } from 'rxjs';
```
Comment (mustaphazorgati) [inline web/src/app/store/Classification-state.ts:2]: Please make the filename lowercase.

File: web/src/app/store/Engine-configuration-state.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
+import { Customisation, WorkbasketsCustomisation, ClassificationsCustomisation, AccessItemsCustomisation, TasksCustomisation, ClassificationCategoryImages } from 'app/models/customisation';
```
Comment (mustaphazorgati) [inline web/src/app/store/Engine-configuration-state.ts:1]: Same as the other state: lowercase filename please.

File: web/src/app/store/Engine-configuration-state.ts [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+import { State, NgxsOnInit, Selector, StateContext, Action } from '@ngxs/store';
```
Comment (mustaphazorgati) [inline web/src/app/store/Engine-configuration-state.ts:2]: We should talk about unit testing our stores.

File: web/src/app/shared/services/classifications/classification-categories.service.ts (hunk 6 above)
Comment (mustaphazorgati) [reply web/src/app/shared/services/classifications/classification-categories.service.ts:14]: => ticket

Comment (mustaphazorgati) [reply web/src/app/shared/services/classifications/classification-categories.service.ts:14]: ich machs.

File: web/src/app/shared/services/classifications/classifications.service.ts (hunk 9 above)
Comment (mustaphazorgati) [reply web/src/app/shared/services/classifications/classifications.service.ts:71]: found out. pls fix

File: web/src/app/shared/services/classifications/classifications.service.ts (hunk 10 above)
Comment (mustaphazorgati) [reply web/src/app/shared/services/classifications/classifications.service.ts:117]: Please refactor the pipe call as discussed - to learn ;)

File: web/src/app/store/Classification-state.ts (hunk 11 above)
Comment (mustaphazorgati) [reply web/src/app/store/Classification-state.ts:1]: https://itnext.io/ngxs-thoughts-patterns-architecture-and-best-practices-c991c42618d9

=> extract selectors + actions

File: web/src/app/administration/classification/details/classification-details.component.ts [hunk 15]
```diff
@@ -166,6 +163,70 @@ export class ClassificationDetailsComponent implements OnInit, OnDestroy {
...
+  validChanged(): void {
+    this.classification.isValidInDomain = !this.classification.isValidInDomain;
+  }
+
+  masterDomainSelected(): boolean {
+    return this.domainService.getSelectedDomainValue() === '';
+  }
+
+  ngOnDestroy(): void {
+    if (this.masterAndDetailSubscription) {
```
[the remark this answers is not in the document]
Comment (sofie29) [reply web/src/app/administration/classification/details/classification-details.component.ts:183]: yes :)
